core/vm/evm.go

```diff
@@ -202,12 +202,13 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 	if !evm.StateDB.Exist(addr) {
 		if !isPrecompile && evm.chainRules.IsEIP4762 {
 			// add proof of absence to witness
-			wgas := evm.Accesses.TouchFullAccount(addr.Bytes(), false)
-			if gas < wgas {
+			gc := gasConsumer{availableGas: gas}
+			ok := evm.Accesses.TouchFullAccount(addr.Bytes(), false, gc.consumeGas)
+			if !ok {
 				evm.StateDB.RevertToSnapshot(snapshot)
 				return nil, 0, ErrOutOfGas
 			}
-			gas -= wgas
+			gas = gc.availableGas
```

Comment:
This is the special case where we use `gasConsumer`. In this `Call(...)` the worst-case gas was already charged. The logic here is trying to use that gas, and later return the `leftOverGas`.

The `UseGas(...)` new callback that we added to the API, is simply this `gasConsumer` struct that allows to us up to `gas`, and we later can see how much was left over and continue as usual.

This is required since we can't do the usual directly consume the contract gas, since it was already charged.